services/QuillLMS/app/services/vitally_integration/serialize_vitally_sales_organization.rb

```diff
@@ -0,0 +1,194 @@
+      last_active_time: last_active_time,
+    }
+  end
+
+  def diagnostic_rollups
+    school_year_start = School.school_year_start(Time.current)
+
+    diagnostics_assigned_this_year = diagnostics_assigned_between_count(school_year_start, school_year_start + 1.year)
+    diagnostics_assigned_last_year = diagnostics_assigned_between_count(school_year_start - 1.year, school_year_start)
+    diagnostics_completed_this_year = diagnostics_completed_between(school_year_start, school_year_start + 1.year)
+    diagnostics_completed_last_year = diagnostics_completed_between(school_year_start - 1.year, school_year_start)
+    percent_completed_this_year = diagnostics_assigned_this_year > 0 ? (1.0 * diagnostics_completed_this_year / diagnostics_assigned_this_year) : 0.0
+    percent_completed_last_year = diagnostics_assigned_last_year > 0 ? (1.0 * diagnostics_completed_last_year / diagnostics_assigned_last_year) : 0.0
+
+    {
+      diagnostics_assigned_this_year: diagnostics_assigned_this_year,
+      diagnostics_assigned_last_year: diagnostics_assigned_last_year,
+      diagnostics_completed_this_year: diagnostics_completed_this_year,
+      diagnostics_completed_last_year: diagnostics_completed_last_year,
+      percent_diagnostics_completed_this_year: percent_completed_this_year,
+      percent_diagnostics_completed_last_year: percent_completed_last_year
+    }
+  end
+
+  def subscription_rollups
+    {
+      premium_start_date: premium_start_date,
+      premium_expiry_date: premium_expiry_date,
+      district_subscription: district_subscription,
+      annual_revenue_current_contract: annual_revenue_current_contract,
+      stripe_invoice_id_current_contract: stripe_invoice_id_current_contract,
+      purchase_order_number_current_contract: purchase_order_number_current_contract
+    }
+  end
+
+  def diagnostics_assigned_between_count(start, stop)
+    @district.schools.select("array_length(classroom_units.assigned_student_ids, 1) AS assigned_students")
+      .joins(users: {
+      classrooms_i_teach: {
+        classroom_units: {
+          unit_activities: {
+           activity: :classification
+          }
+        }
+      }
+    }).where(classification: {key: ActivityClassification::DIAGNOSTIC_KEY})
+      .where(classroom_units: {created_at: start..stop})
+      .map(&:assigned_students).reject(&:blank?).sum
+  end
+
+  def diagnostics_completed_between(start, stop)
+    @district.schools.joins(users: {
+      classrooms_i_teach: {
+        classroom_units: {
+          activity_sessions: {
+            activity: :classification
+          }
+        }
+      }
+    }).where(classification: {key: ActivityClassification::DIAGNOSTIC_KEY})
+      .where(activity_sessions: {completed_at: start..stop})
+      .distinct
+      .count
+  end
+
+  def active_students(start_date=nil, end_date=nil)
+    return activities_completed_all_time.group("students.id").length if start_date.blank? && end_date.blank?
```

Comment:
this is a newly added method.

Reply:
I think that `count` is preferable to `length` here.  ActiveRecord runs the `COUNT(*)` command in SQL, but I'm pretty sure that `length` pulls all items into Rails and the counts the length of the array which can be way less efficient.